File: spring-data/src/main/java/dev/cerbos/queryplan/springdata/CollectionTranslator.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.queryplan.springdata;

import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter;
import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter.Expression.Operand;

import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.Expression;
import jakarta.persistence.criteria.Predicate;

import com.google.protobuf.Value;

import java.util.List;
import java.util.Set;

/**
 * Translates {@code exists}, {@code exists_one} and {@code all} over a relation chain, and
 * folds them over a literal list. The subqueries come from {@link ChainSubqueries}. {@code filter} in boolean position is refused by the walk, and
 * {@code size(filter(...))} belongs to {@link SizeTranslator}.
 */
final class CollectionTranslator {

    private static final Set<String> LAMBDA_BINDING_OPERATORS =
            Set.of("exists", "exists_one", "all", "filter", "map", "except");

    private final CriteriaBuilder cb;
    private final PlanWalker walker;
    private final ChainSubqueries subqueries;

    CollectionTranslator(CriteriaBuilder cb, PlanWalker walker, ChainSubqueries subqueries) {
        this.cb = cb;
        this.walker = walker;
        this.subqueries = subqueries;
    }

    /**
     * Translates a macro over a relation as a tri-state predicate. A NULL element column is a
     * missing attribute, so a body that reads it is a CEL error, mapped to SQL UNKNOWN:
     * <ul>
     *   <li>{@code exists}: true if any element matches, else UNKNOWN if any errors, else
     *       false;</li>
     *   <li>{@code all}: false if any element fails, else UNKNOWN if any errors, else
     *       true;</li>
     *   <li>{@code exists_one}: UNKNOWN if any element errors, else true if exactly one
     *       matches.</li>
     * </ul>
     * A plain EXISTS would turn an error into FALSE, and its negation would return rows the
     * PDP denies.
     *
     * <p>The body is translated twice per macro (three times for {@code exists_one}), so
     * nesting multiplies subqueries. {@link PlanWalker#enterMacro} bounds the depth
     * ({@link SpringDataQueryPlanAdapter#MAX_MACRO_DEPTH_PROPERTY}, default
     * {@value SpringDataQueryPlanAdapter#DEFAULT_MAX_MACRO_DEPTH}).
     */
    Predicate handleCollectionOperator(String op, List<Operand> operands, Scope scope) {
        if (operands.size() != 2) {
            throw Refusals.malformed(op + " requires exactly 2 operands");
        }
        Operand listOperand = operands.get(0);
        Operand lambdaOperand = operands.get(1);

        // The planner unrolls a macro over a known list of up to 10 elements itself; above
        // that the list arrives as the collection operand, so fold it here the same way.
        if (listOperand.getNodeCase() == Operand.NodeCase.VALUE) {
            return walker.enterMacro(op,
                    () -> handleKnownValueCollection(op, listOperand.getValue(), lambdaOperand, scope));
        }

        if (listOperand.getNodeCase() != Operand.NodeCase.VARIABLE) {
            // e.g. `tags.map(...).exists(...)`: legal CEL with no join chain.
            throw Refusals.unsupported(op + " first operand must be a variable");
        }
        if (lambdaOperand.getNodeCase() != Operand.NodeCase.EXPRESSION
                || !"lambda".equals(lambdaOperand.getExpression().getOperator())) {
            throw Refusals.malformed(op + " second operand must be a lambda");
        }

        String collectionVar = listOperand.getVariable();
        if (!(scope.resolve(collectionVar) instanceof Scope.ResolvedRelation ref)) {
            throw Refusals.unmapped(
                    op + " requires a Relation mapping for " + collectionVar);
        }

        ParsedLambda lambda = ParsedLambda.parse(lambdaOperand,
                op + " second operand must be a lambda",
                "lambda requires exactly 2 operands",
                "lambda variable must be a variable operand");
        Operand body = lambda.body();
        String lambdaVarName = lambda.varName();

        // Each call builds a fresh Predicate tree: Hibernate 6 negation is stateful.
        SubqueryBodyBuilder bodyBuilder = (sub, tailJoin, rebased) -> walker.traverse(body,
                Scope.lambda(tailJoin, sub, ref.tail(), lambdaVarName, rebased));

        return walker.enterMacro(op, () -> switch (op) {
            case "exists" ->
                    cb.equal(subqueries.requireLeadingHops(scope, ref,
                            subqueries.macroScoreSubquery(scope, ref, bodyBuilder, 2, 0),
                            Integer.class), 2);
            case "all" ->
                    cb.equal(subqueries.requireLeadingHops(scope, ref,
                            subqueries.macroScoreSubquery(scope, ref, bodyBuilder, 0, 2),
                            Integer.class), 0);
            case "exists_one" ->
                    cb.equal(subqueries.requireLeadingHops(scope, ref,
                            subqueries.strictMatchCountSubquery(scope, ref, bodyBuilder),
                            Long.class), 1L);
            default -> throw Refusals.internal("Unsupported collection operator: " + op);
        });
    }

    /**
     * Folds a macro over a literal list: substitutes each element into the body and joins the
     * results with {@code or} ({@code exists}) or {@code and} ({@code all}), as the planner
     * does for short lists. An empty list gives false for {@code exists} and true for
     * {@code all}. {@code exists_one} is the strict count ({@link #literalStrictCount}) equal
     * to 1.
     */
    private Predicate handleKnownValueCollection(String op, Value collectionValue,
                                                 Operand lambdaOperand, Scope scope) {
        if (!"exists".equals(op) && !"all".equals(op) && !"exists_one".equals(op)) {
            throw Refusals.unsupported(op
                    + " over a literal collection value is not supported. "
                    + "Only exists(), all() and exists_one() can be folded into a flat filter.");
        }
        if (collectionValue.getKindCase() != Value.KindCase.LIST_VALUE) {
            // CEL cannot iterate a scalar, so the planner cannot emit this.
            throw Refusals.malformed(op
                    + " over a literal collection requires a list value");
        }
        ParsedLambda lambda = ParsedLambda.parse(lambdaOperand,
                op + " second operand must be a lambda",
                op + " over a literal collection supports single-variable lambdas only",
                "lambda variable must be a variable operand");

        if ("exists_one".equals(op)) {
            return cb.equal(literalStrictCount(cb, walker, collectionValue, lambda, scope), 1L);
        }
        List<Value> elements = collectionValue.getListValue().getValuesList();
        if (elements.isEmpty()) {
            return "exists".equals(op) ? cb.disjunction() : cb.conjunction();
        }

        PlanResourcesFilter.Expression.Builder combined = PlanResourcesFilter.Expression
                .newBuilder()
                .setOperator("exists".equals(op) ? "or" : "and");
        for (Value element : elements) {
            combined.addOperands(
                    substituteLambdaVariable(lambda.body(), lambda.varName(), element));
        }
        return walker.traverse(Operand.newBuilder().setExpression(combined).build(), scope);
    }

    /** A lambda variable no plan names, bound to the element a positional read selects. */
    private static final String POSITIONAL_ELEMENT = "__cerbos_positional_element";

    /**
     * A leaf that reads one list element by position, {@code R.attr.list[i] op v} or
     * {@code R.attr.list[i].field op v}, over a relation that declares its
     * {@linkplain AttributeMapping.Relation#withPositionField position field}. Returns
     * {@code null} when no operand is such a read, so the leaf is translated as usual.
     *
     * <p>It is translated as {@code list.exists(e, position(e) == i && e op v)}, whose score
     * subquery is TRUE when the element at {@code i} satisfies the leaf, FALSE when it fails
     * it, and UNKNOWN when its comparison is (every other element is FALSE). CEL errors when
     * {@code i} is not a non-negative integer or no element sits at {@code i}, so those are
     * UNKNOWN too, under both polarities.
     */
    Predicate tryPositionalRead(String op, List<Operand> operands, Scope scope) {
        int readAt = -1;
        for (int i = 0; i < operands.size(); i++) {
            if (positionalRead(operands.get(i)) != null) {
                if (readAt >= 0) {
                    throw Refusals.unsupported(op + " between two positional list reads is not"
                            + " supported: each is a separate correlated element");
                }
                readAt = i;
            }
        }
        if (readAt < 0) {
            return null;
        }
        PlanResourcesFilter.Expression index = positionalRead(operands.get(readAt));
        Operand read = operands.get(readAt);
        String field = read.getExpression().getOperator().equals("get-field")
                ? read.getExpression().getOperands(1).getVariable() : null;
        String listVar = index.getOperands(0).getVariable();
        if (!(scope.resolve(listVar) instanceof Scope.ResolvedRelation ref)
                || ref.isChained() || ref.tail().positionField() == null) {
            throw Refusals.unsupported("Positional access into " + listVar + " requires a"
                    + " direct Relation mapping that declares its position field"
                    + " (AttributeMapping.Relation#withPositionField): a JPA collection has no"
                    + " order a plan can name otherwise");
        }
        TriPredicate tri = new TriPredicate(cb);
        Value position = index.getOperands(1).getValue();
        double at = position.getKindCase() == Value.KindCase.NUMBER_VALUE
                ? position.getNumberValue() : -1;
        if (at < 0 || at != Math.rint(at) || at > Integer.MAX_VALUE) {
            // A non-integral, negative or non-numeric index is a CEL error.
            return tri.unknown();
        }
        int atIndex = (int) at;
        String positionField = ref.tail().positionField();

        Operand element = Operand.newBuilder().setVariable(field == null
                ? POSITIONAL_ELEMENT : POSITIONAL_ELEMENT + "." + field).build();
        PlanResourcesFilter.Expression.Builder leaf = PlanResourcesFilter.Expression.newBuilder()
                .setOperator(op).addAllOperands(operands).setOperands(readAt, element);
        Operand body = Operand.newBuilder().setExpression(leaf).build();
        SubqueryBodyBuilder bodyBuilder = (sub, tailJoin, rebased) -> cb.and(
                cb.equal(tailJoin.get(positionField), atIndex),
                walker.traverse(body, Scope.lambda(tailJoin, sub, ref.tail(),
                        POSITIONAL_ELEMENT, rebased)));
        return walker.enterMacro(op, () -> tri.baseUnlessUnknown(
                cb.equal(subqueries.macroScoreSubquery(scope, ref, bodyBuilder, 2, 0), 2),
                () -> tri.not(subqueries.existsSubquery(scope, ref, (sub, tailJoin, rebased) ->
                        cb.equal(tailJoin.get(positionField), atIndex)))));
    }

    /** The {@code index(list, i)} {@code operand} reads, directly or under a get-field. */
    private static PlanResourcesFilter.Expression positionalRead(Operand operand) {
        if (operand.getNodeCase() != Operand.NodeCase.EXPRESSION) {
            return null;
        }
        PlanResourcesFilter.Expression e = operand.getExpression();
        if ("get-field".equals(e.getOperator()) && e.getOperandsCount() == 2
                && e.getOperands(1).getNodeCase() == Operand.NodeCase.VARIABLE) {
            PlanResourcesFilter.Expression inner = positionalRead(e.getOperands(0));
            return inner != null && "index".equals(e.getOperands(0).getExpression().getOperator())
                    ? inner : null;
        }
        return "index".equals(e.getOperator()) && e.getOperandsCount() == 2
                && e.getOperands(0).getNodeCase() == Operand.NodeCase.VARIABLE
                && e.getOperands(1).getNodeCase() == Operand.NodeCase.VALUE ? e : null;
    }

    /**
     * The number of elements of a literal list whose body holds, or NULL when any body is
     * UNKNOWN: {@code exists_one} and {@code filter} evaluate every element and error if any
     * does. Each element contributes {@code CASE WHEN p THEN 1 WHEN NOT p THEN 0 END}, which is
     * NULL for an UNKNOWN body, and SQL {@code +} carries that NULL into the sum.
     */
    static Expression<Long> literalStrictCount(CriteriaBuilder cb, PlanWalker walker,
                                               Value list, ParsedLambda lambda, Scope scope) {
        if (list.getKindCase() != Value.KindCase.LIST_VALUE) {
            // CEL cannot iterate a scalar, so the planner cannot emit this.
            throw Refusals.malformed("a macro over a literal collection requires a list value");
        }
        TriPredicate tri = new TriPredicate(cb);
        Expression<Long> count = cb.literal(0L);
        for (Value element : list.getListValue().getValuesList()) {
            Operand body = substituteLambdaVariable(lambda.body(), lambda.varName(), element);
            // Built twice: a predicate node must not be used in both polarities.
            Expression<Long> matched = cb.<Long>selectCase()
                    .when(walker.traverse(body, scope), 1L)
                    .when(tri.not(walker.traverse(body, scope)), 0L)
                    .otherwise(cb.nullLiteral(Long.class));
            count = cb.sum(count, matched);
        }
        return count;
    }

    /**
     * Replaces the lambda variable with {@code element} in a body. {@code var.a.b} reads a
     * field of the element. A nested macro that rebinds the same name shadows it, so only its
     * collection operand is substituted.
     */
    private static Operand substituteLambdaVariable(Operand operand, String varName,
                                                    Value element) {
        switch (operand.getNodeCase()) {
            case VARIABLE -> {
                String name = operand.getVariable();
                if (name.equals(varName)) {
                    return Operand.newBuilder().setValue(element).build();
                }
                if (name.startsWith(varName + ".")) {
                    return Operand.newBuilder()
                            .setValue(resolveElementPath(name,
                                    name.substring(varName.length() + 1), element))
                            .build();
                }
                return operand;
            }
            case EXPRESSION -> {
                PlanResourcesFilter.Expression expr = operand.getExpression();
                List<Operand> ops = expr.getOperandsList();
                PlanResourcesFilter.Expression.Builder rebuilt = expr.toBuilder();
                if (LAMBDA_BINDING_OPERATORS.contains(expr.getOperator()) && ops.size() == 2
                        && shadowsVariable(ops.get(1), varName)) {
                    rebuilt.setOperands(0,
                            substituteLambdaVariable(ops.get(0), varName, element));
                    return Operand.newBuilder().setExpression(rebuilt).build();
                }
                for (int i = 0; i < ops.size(); i++) {
                    rebuilt.setOperands(i,
                            substituteLambdaVariable(ops.get(i), varName, element));
                }
                return Operand.newBuilder().setExpression(rebuilt).build();
            }
            default -> {
                return operand;
            }
        }
    }

    private static boolean shadowsVariable(Operand lambdaOperand, String varName) {
        if (lambdaOperand.getNodeCase() != Operand.NodeCase.EXPRESSION
                || !"lambda".equals(lambdaOperand.getExpression().getOperator())) {
            return false;
        }
        List<Operand> ops = lambdaOperand.getExpression().getOperandsList();
        return ops.size() == 2
                && ops.get(1).getNodeCase() == Operand.NodeCase.VARIABLE
                && varName.equals(ops.get(1).getVariable());
    }

    private static Value resolveElementPath(String fullRef, String path, Value element) {
        Value current = element;
        for (String segment : path.split("\\.")) {
            if (current.getKindCase() != Value.KindCase.STRUCT_VALUE
                    || !current.getStructValue().containsFields(segment)) {
                // CEL would error on this element, and the fold cannot express a
                // per-element error, so refuse.
                throw Refusals.unsupported("Cannot resolve \"" + fullRef
                        + "\": collection element has no field \"" + segment + "\"");
            }
            current = current.getStructValue().getFieldsOrThrow(segment);
        }
        return current;
    }
}

File: exposed/src/main/kotlin/dev/cerbos/queryplan/exposed/CollectionTranslator.kt
package dev.cerbos.queryplan.exposed

import com.google.protobuf.Value
import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter
import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter.Expression.Operand
import dev.cerbos.queryplan.exposed.sql.ScoreCase
import dev.cerbos.queryplan.exposed.sql.nullIfUndetermined
import org.jetbrains.exposed.v1.core.Alias
import org.jetbrains.exposed.v1.core.Coalesce
import org.jetbrains.exposed.v1.core.EqOp
import org.jetbrains.exposed.v1.core.IntegerColumnType
import org.jetbrains.exposed.v1.core.Max
import org.jetbrains.exposed.v1.core.Op
import org.jetbrains.exposed.v1.core.PlusOp
import org.jetbrains.exposed.v1.core.Sum
import org.jetbrains.exposed.v1.core.Table
import org.jetbrains.exposed.v1.core.intLiteral

/**
 * The collection macros `exists`, `all` and `exists_one`, lowered to ONE scoring subquery each,
 * never a plain `EXISTS`: an element whose body is UNKNOWN silently fails to match under `EXISTS`,
 * which collapses CEL's error case to FALSE, and an enclosing NOT then readmits the row.
 *
 * The truth tables the scores encode are CEL's, not SQL's:
 *  - `exists` is OR with error absorption — TRUE if any element matches, ERROR if none matches and
 *    at least one errors, FALSE otherwise (including the empty collection);
 *  - `all` is AND with error absorption — FALSE if any element fails, ERROR if none fails and at
 *    least one errors, TRUE otherwise (including the empty collection);
 *  - `exists_one` is STRICT — any erroring element errors the whole macro, otherwise exactly one
 *    match.
 *
 * ERROR becomes SQL UNKNOWN, so the row stays excluded under both polarities.
 */
internal class CollectionTranslator(private val translation: Translation) {

    fun translate(operator: String, operands: List<Operand>, scope: Scope): Op<Boolean> {
        if (operands.size != 2) {
            throw Refusals.malformed("$operator requires exactly 2 operands, got ${operands.size}")
        }
        val collectionOperand = operands[0]
        val lambdaOperand = operands[1]

        // A literal value-list collection arrives when the planner could not unroll the macro
        // itself: at 10 elements or fewer it folds exists/all into an or/and chain, above that the
        // lambda ships with the folded list as its collection operand. Applying the same fold here
        // is what keeps the translation independent of which side of that threshold it landed on.
        //
        // It costs a macro level like any other. The fold emits no subquery of its own, but it
        // DUPLICATES everything under it once per element, so two three-element folds over one
        // relation macro emit nine correlated subqueries and an eleven-element pair emits 121.
        // Reaching it before [PlanWalker.enterMacro] left the emitted expression unbounded whatever
        // [Options.maxMacroDepth] was set to.
        if (collectionOperand.nodeCase == Operand.NodeCase.VALUE) {
            return translation.walker.enterMacro(operator) {
                foldKnownValues(operator, collectionOperand.value, lambdaOperand, scope)
            }
        }
        if (collectionOperand.nodeCase != Operand.NodeCase.VARIABLE) {
            throw RelationRefusals.computedCollection(operator)
        }
        val collection = scope.collection(collectionOperand.variable)
        val lambda = ParsedLambda.parse(
            lambdaOperand,
            "$operator second operand must be a lambda",
            "$operator supports single-variable lambdas only",
            "$operator lambda variable must be a variable operand",
        )

        return translation.walker.enterMacro(operator) {
            when (operator) {
                // The dominant score with the empty collection folded to 0 and "only undetermined
                // elements" (max score 1) mapped to SQL NULL. A two-valued equality against the
                // witness score then yields TRUE / FALSE / UNKNOWN exactly per the CEL table.
                "exists" -> scored(collection, lambda, scope, trueScore = 2, falseScore = 0, witness = 2)
                "all" -> scored(collection, lambda, scope, trueScore = 0, falseScore = 2, witness = 0)
                "exists_one" -> strictlyOne(collection, lambda, scope)
                else -> throw Refusals.internal("Unsupported collection operator: $operator")
            }
        }
    }

    /**
     * `NULLIF(COALESCE(MAX(CASE WHEN body THEN t WHEN NOT body THEN f ELSE 1 END), 0), 1) = witness`.
     *
     * One scan, one pair of scores. The scores are chosen so the macro's own witness state
     * dominates the MAX: 2 for the state that decides the macro outright, 0 for its opposite, and
     * 1 for undetermined — which therefore only wins when no element decided anything, and NULLIF
     * turns that win into the UNKNOWN the row has to inherit.
     */
    private fun scored(
        collection: Resolution.Collection,
        lambda: ParsedLambda,
        scope: Scope,
        trueScore: Int,
        falseScore: Int,
        witness: Int,
    ): Op<Boolean> {
        val fold = translation.subqueries.chainAggregate(collection, IntegerColumnType()) { alias ->
            val body = translateBody(collection, lambda, scope, alias)
            nullIfUndetermined(
                Coalesce<Int, Int?>(
                    Max<Int, Int>(ScoreCase(body, trueScore, falseScore, 1), IntegerColumnType()),
                    intLiteral(0),
                ),
            )
        }
        return EqOp(fold, intLiteral(witness))
    }

    /**
     * `COALESCE(SUM(CASE WHEN body THEN 1 ELSE 0 END), 0) + <poison> = 1`.
     *
     * `exists_one` has no error absorption: one undetermined element errors the whole macro even
     * when the count would otherwise hold. The poison term is 0 when every element is determined
     * and SQL NULL otherwise, and NULL is absorbing under addition, so the comparison goes UNKNOWN
     * exactly when CEL errors. The empty collection yields 0 + 0 = 0, which compares FALSE — which
     * is what CEL's `exists_one` over an empty list is.
     */
    private fun strictlyOne(
        collection: Resolution.Collection,
        lambda: ParsedLambda,
        scope: Scope,
    ): Op<Boolean> {
        val total = translation.subqueries.chainAggregate(collection, IntegerColumnType()) { alias ->
            val body = translateBody(collection, lambda, scope, alias)
            val matches = Coalesce<Int, Int?>(
                Sum<Int>(ScoreCase(body, 1, 0, 0), IntegerColumnType()),
                intLiteral(0),
            )
            val poison = nullIfUndetermined(
                Coalesce<Int, Int?>(
                    Max<Int, Int>(ScoreCase(body, 0, 0, 1), IntegerColumnType()),
                    intLiteral(0),
                ),
            )
            PlusOp<Int, Int>(matches, poison, IntegerColumnType())
        }
        return EqOp(total, intLiteral(1))
    }

    /**
     * The lambda body, translated against the element the subquery ranges over.
     *
     * The element scope delegates every non-lambda variable outward, so an outer reference inside
     * the body resolves against the scope the macro appeared in and reaches the subquery as an
     * ordinary correlation. The body is built ONCE and shared between the polarities a score needs:
     * an Exposed expression tree is immutable, so a node is safe to appear twice.
     */
    private fun translateBody(
        collection: Resolution.Collection,
        lambda: ParsedLambda,
        scope: Scope,
        alias: Alias<Table>,
    ): Op<Boolean> = translation.walker.traverse(
        lambda.body,
        LambdaScope(translation, alias, collection.tail, lambda.variable, scope),
    )

    /**
     * Fold a macro whose collection operand is a literal value list: substitute each element into
     * the lambda body and walk the combined `or`/`and` expression, which is the same fold the
     * planner applies below its own unroll threshold.
     *
     * Element comparisons produced by the fold flow through the ordinary comparison translation, so
     * a NULL column keeps its SQL-UNKNOWN reading exactly as it does in a chain the planner
     * unrolled itself.
     */
    private fun foldKnownValues(
        operator: String,
        collection: Value,
        lambdaOperand: Operand,
        scope: Scope,
    ): Op<Boolean> {
        if (operator != "exists" && operator != "all") {
            throw RelationRefusals.unfoldableValueCollection(operator)
        }
        if (collection.kindCase != Value.KindCase.LIST_VALUE) {
            // CEL refuses a scalar as a comprehension range, so the planner never folds one.
            throw Refusals.malformed("$operator over a literal collection requires a list value")
        }
        val lambda = ParsedLambda.parse(
            lambdaOperand,
            "$operator second operand must be a lambda",
            "$operator over a literal collection supports single-variable lambdas only",
            "$operator lambda variable must be a variable operand",
        )
        val elements = collection.listValue.valuesList
        // CEL identity over the empty collection: exists() matches nothing, all() matches
        // everything. Both are DEFINITE, so a constant is the whole answer.
        if (elements.isEmpty()) {
            return if (operator == "exists") Op.FALSE else Op.TRUE
        }
        val combined = PlanResourcesFilter.Expression.newBuilder()
            .setOperator(if (operator == "exists") "or" else "and")
        elements.forEach { element ->
            combined.addOperands(substitute(lambda.body, lambda.variable, element))
        }
        return translation.walker.traverse(Operand.newBuilder().setExpression(combined).build(), scope)
    }

    /**
     * Substitute a lambda's iteration variable with one concrete element.
     *
     * A bare reference becomes the element; a dotted one drills into it. A nested macro that
     * REBINDS the same variable name shadows this one, so substitution only descends into its
     * collection operand — otherwise the inner macro's own elements would be replaced by the outer
     * fold's.
     */
    private fun substitute(operand: Operand, variable: String, element: Value): Operand =
        when (operand.nodeCase) {
            Operand.NodeCase.VARIABLE -> {
                val name = operand.variable
                when {
                    name == variable -> Operand.newBuilder().setValue(element).build()
                    name.startsWith("$variable.") ->
                        Operand.newBuilder()
                            .setValue(drill(name, name.substring(variable.length + 1), element))
                            .build()
                    else -> operand
                }
            }
            Operand.NodeCase.EXPRESSION -> {
                val expression = operand.expression
                val rebuilt = expression.toBuilder()
                val operands = expression.operandsList
                if (expression.operator in LAMBDA_BINDING_OPERATORS &&
                    operands.size == 2 &&
                    shadows(operands[1], variable)
                ) {
                    rebuilt.setOperands(0, substitute(operands[0], variable, element))
                } else {
                    operands.forEachIndexed { index, nested ->
                        rebuilt.setOperands(index, substitute(nested, variable, element))
                    }
                }
                Operand.newBuilder().setExpression(rebuilt).build()
            }
            else -> operand
        }

    private fun shadows(lambdaOperand: Operand, variable: String): Boolean {
        if (lambdaOperand.nodeCase != Operand.NodeCase.EXPRESSION ||
            lambdaOperand.expression.operator != "lambda"
        ) {
            return false
        }
        val operands = lambdaOperand.expression.operandsList
        return operands.size == 2 &&
            operands[1].nodeCase == Operand.NodeCase.VARIABLE &&
            operands[1].variable == variable
    }

    private fun drill(reference: String, path: String, element: Value): Value {
        var current = element
        path.split('.').forEach { segment ->
            if (current.kindCase != Value.KindCase.STRUCT_VALUE ||
                !current.structValue.containsFields(segment)
            ) {
                throw RelationRefusals.unresolvableFoldedElement(reference, segment)
            }
            current = current.structValue.getFieldsOrThrow(segment)
        }
        return current
    }

    private companion object {
        /** Operators whose second operand is a lambda that binds an iteration variable. */
        val LAMBDA_BINDING_OPERATORS = setOf("exists", "exists_one", "all", "filter", "map", "except")
    }
}
